manticore/ethereum.py

```diff
@@ -1078,6 +1123,38 @@ def global_findings(self):
     def workspace(self):
         return self._executor._workspace._store.uri
 
+    def _concretize_offsets_and_sizes(self, state, signature, data):
+        ''' Using signature spec this function browse the data and concretize 
+            all the offsets to dynamic arguments and all the size of the dynamic
+            arguments so it all fits into data.
+            The available space is fairly divided among all the dynamic arguments.
+        '''
+        is_multiple, func_name, types = ABI.parse_signature_spec(signature)
+        dyn_arguments = []
+        for pos, t in enumerate(types):
+            dynamic = t in (u'bytes', u'string') or t.endswith(']')
+            if dynamic:
+                dyn_arguments.append(pos)
+
+        used = 4 + len(types)*32 
+        available = len(data) -used -len(dyn_arguments)*32
+
+        for i in range(0, len(dyn_arguments)):
+            #get, constraint and concretize dyn_offset to some reasonable value
+            dyn_offset = ABI.get_uint(data, 4 + dyn_arguments[i]*32, 256)
+            assert solver.can_be_true(state.constraints, dyn_offset == used)
+            state.constrain(dyn_offset == used)
+            data[used:used+32] = ("%064x"%(available/32/len(dyn_arguments))).decode('hex')
+
+            #get, constraint and concretize dyn_size to some reasonable value
+            dyn_size = ABI.get_uint(data, used, 256)
```

Comment:
i don't understand this block. i thought the above block concretizes the offset and the size, and then we're done for this argument?

Reply:
First you add a constraint to the path constrraint. 
constraints.add(data[10:20] == 'AAAA..AA')
Then you overwrite the state (data) with the concrete value.
data[10:20] = 'AAAA..AA'
So next time you read data[10:20] you get a constant value and not a symbolic variable with a single solution. ?

Reply:
sorry for being a bit slow, trying to wrap my head around concretizing a variable :x 😬 

would it be correct to say that overwriting the state with the concrete value is an optimization? and technically all that is necessary for _correctness_ is to constrain the variable? not saying we shouldn't do it, but just checking

Reply:
Yes, writing the concrete value to the state (after constraining it to a single value) is somehow redundant.